### ISSUE TITLE
Where should my ontology go? Data graph versus shapes graph

### ISSUE THREAD

**Issue by wouterbeek** [w3c/data-shapes#185]
Originally posed over at <https://github.com/w3c/data-shapes/issues/155>; also see the comments by others over there.

### Observation

According to the SHACL standard, two graphs are relevant for validation: the data graph and the shapes graph. The ontology should be part of the **data graph**:

> The data graph is expected to include all the ontology axioms related to the data and especially all the rdfs:subClassOf triples in order for SHACL to correctly identify class targets and validate Core SHACL constraints.

This seems counter-intuitive to me, since I associate the ontology more with the shapes graph. For example, a shapes graph can `owl:import` an ontology.

### Example

To illustrate my unease, let's take the following **data graph**:

```turtle
prefix id: <https://example.com/>
prefix foaf: <http://xmlns.com/foaf/0.1/>

id:john a foaf:Person.
```

And the following **shapes graph**:

```turtle
prefix foaf: <http://xmlns.com/foaf/0.1/>
prefix sh: <http://www.w3.org/ns/shacl#>

[] sh:targetClass foaf:Agent;
   sh:property
     [ sh:minCount 1;
       sh:path foaf:name ].
```

Adding the following **ontology graph** is crucial, otherwise we cannot invalidate the data graph which is missing a `foaf:name` statement:

```turtle
prefix foaf: <http://xmlns.com/foaf/0.1/>
prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>

foaf:Person rdfs:subClassOf foaf:Agent.
```

### Use case

I have a specific use case where this comes up: in TriplyETL we stream though the instance data. The stream passes along millions of small data graphs. For each of these data graphs, we have to add the ontology before the data graph can be validated in-stream. In this use case, it makes more sense to add the ontology to the shapes graph once, and use that same shapes graph to validate all data graphs that pass by.

### Expected

I expect either of the following:

- The SHACL standard requires that the ontology graph is added to the shapes graph, and not to the data graph.
- The SHACL standard requires that the ontology graph is added to the data graph and/or the shapes graph.

**Comment by HolgerKnublauch** [comment]
I guess rdfs:subClassOf triples are what matters here. They impact sh:class and class-based targets.

I believe we could change SHACL Core so that these triples will be considered from the union of data and shapes graphs.

Would this address your concern or are there other triples in the data graph that should also be in the shapes graph and vice versa?

**Comment by bergos** [comment]
Can't we define that all `rdfs:subClassOf` reasoning must happen in the shapes graph?

A union graph could make some edge cases, like validating constraints on a SHACL shape, difficult to process. A flag to enable that feature could solve it, but we should only consider it if there are other use cases than the `rdfs:subClassOf` reasoning.

**Comment by HolgerKnublauch** [comment]
If we were to ignore rdfs:subClassOf triples from the data graph then we would introduce a breaking change to SHACL, which is something we definitely want to avoid for this (incremental) release. Adding the shapes graph as an extra graph to process is less likely to break existing use cases. But even that is potentially breaking.

**Comment by wouterbeek** [comment]
An alternative solution is to introduce a new 3rd graph:

1. **Data graph**, containing instances of the classes defined in (2) and (3).
2. **Shapes graph**, containing node shapes and property shapes for the instance data in (1). This is the closed half of the data model (SHACL).
3. **Ontology graph**, optionally containing classes and properties for the instance data in (1). This is the open half of the data model (RDFS/OWL).

In SHACL 1.0 graph 3 is never given.

In SHACL 1.1, it becomes possible to optionally specify graph 3. If graph 3 is specified, then all class and property statements (RDFS/OWL, including `rdfs:subClassOf`) are assumed to be in that graph. A user can choose to specify the same graph for (2) and (3).

**Comment by VladimirAlexiev** [comment]
Related: https://github.com/w3c/data-shapes/issues/183

**Comment by ajnelson-nist** [comment]
> An alternative solution is to introduce a new 3rd graph:
> 
> 1. **Data graph**, containing instances of the classes defined in (2) and (3).
> 2. **Shapes graph**, containing node shapes and property shapes for the instance data in (1). This is the closed half of the data model (SHACL).
> 3. **Ontology graph**, optionally containing classes and properties for the instance data in (1). This is the open half of the data model (RDFS/OWL).
> 
> In SHACL 1.0 graph 3 is never given.
> 
> In SHACL 1.1, it becomes possible to optionally specify graph 3. If graph 3 is specified, then all class and property statements (RDFS/OWL, including `rdfs:subClassOf`) are assumed to be in that graph. A user can choose to specify the same graph for (2) and (3).

FWIW, I'm aware of at least one tool that follows this practice, keeping data and ontology graphs separate as tool inputs but mixing them in in-memory.

On the other hand, I work with an ontology community that uses shapes as part of its ontology specification, keeping the ontology graph and shapes graph together.  This makes significant use of [Implicit Class Targets](https://www.w3.org/TR/2017/REC-shacl-20170720/#implicit-targetClass) in co-typing `sh:NodeShape` and `owl:Class`.

That community happens to use that tool I noted, so the end result is two graphs (shapes graph S, and ontology graph O) reviewing three (shapes graph S, ontology graph O, and data graph D - and yes, S=S and O=O).

I just wanted to leave this user story as "anecdata," which might or might not help @wouterbeek 's original guessed either-or:

> I expect either of the following:
>
> * The SHACL standard requires that the ontology graph is added to the shapes graph, and not to the data graph.
> * The SHACL standard requires that the ontology graph is added to the data graph and/or the shapes graph.

I think we will learn the right way forward more from the inferencing work.  My hunch is that the ontology and data graphs will typically have different update rhythms.  If data updates, the ontology graph probably(?) wouldn't need to re-run inferencing.  If the ontology graph updates, the data graph would probably need to re-run inferencing.

**Comment by VladimirAlexiev** [comment]
@ajnelson-nist 
> If data updates, the ontology graph probably(?) wouldn't need to re-run inferencing.
> If the ontology graph updates, the data graph would probably need to re-run inferencing.

Both Data (A-Box) and ontology (T-Box) are needed to have any inferencing.
Given some family ontology, if I add `X ex:son Y` to the data, that should infer `X ex:child Y`.
Not sure what you mean by "re-run" but there are 3 major kinds of inference:
- backward chaining, at query time, eg by expanding a query `?x ex:child ?y` to also look for `ex:son`
- forward chaining, incremental: when a triple is added/deleted, only relevant rules are rerun to add/delete inferences
- forward chaining, full: you periodically need to rerun inference in full

Eg GraphDB is incremental. A major strength is that you can add either Data or Ontology triples **at any time**, and only the needed inference will be rerun.

**Comment by HolgerKnublauch** [comment]
I still want to resolve this issue.

I agree with the original problem statement. In particular, it should be possible to tell the SHACL engine that it should look up the rdfs:subClassOf triples in the shapes graph. This could be a simple boolean flag, which defaults to false. If true then the definition of SHACL Instance and related terms would change to only use the subClassOf triples in the shapes graph.

I will try to work on a PR in this direction.

**Comment by afs** [comment]
re: PR #447 

The "3 graphs" approach (https://github.com/w3c/data-shapes/issues/185#issuecomment-2602019596) seems to handle this. If it is the shapes graph, then it is this PR; if it is the data graph its SHACL 1.0; if it is some of all of the data graph (classes etc for this specific data area) and business-wide ontology, even top-level ontology, then that can be assembled.

The key is "where/how" to do lookups of subclassof. The "how" it is structured can be outside the standard.

At least the shapes graph can have imports, but the team for shapes for a data product and team for a business ontology could be different teams.

**Comment by HolgerKnublauch** [comment]
Yes, my thinking was that it is easier for a user to control what goes into the Shapes Graph. For example that could be a union graph. However, it is usually difficult to control what the data graph is. Therefore I consider it sufficient to support a flag for the shapes graph, as that could include the ontology graph.
